File: flink-java-core/src/main/java/com/atguigu/flink/day3/MinMaxTempAggWindow.java
package com.atguigu.flink.day3;

import com.atguigu.flink.day2.SensorReading;
import com.atguigu.flink.day2.SensorSource;
import org.apache.flink.api.common.functions.AggregateFunction;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.windowing.ProcessWindowFunction;
import org.apache.flink.streaming.api.windowing.time.Time;
import org.apache.flink.streaming.api.windowing.windows.TimeWindow;
import org.apache.flink.util.Collector;

/**
 * @author ：hmz
 * @date ：Created in 2020/9/14 8:37
 */
public class MinMaxTempAggWindow {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setParallelism(1);

        env.addSource(new SensorSource())
                .keyBy(r -> r.getId())
                .timeWindow(Time.seconds(5))
                .aggregate(new Agg(), new Win())
                .print();

        env.execute();
    }

    public static class Agg implements AggregateFunction<SensorReading, Tuple2<Double, Double>, Tuple2<Double, Double>> {

        @Override
        public Tuple2<Double, Double> createAccumulator() {
            return Tuple2.of(Double.MAX_VALUE, Double.MIN_VALUE);
        }

        @Override
        public Tuple2<Double, Double> add(SensorReading value, Tuple2<Double, Double> accumulator) {
            accumulator.f0.min(value.getTamperature(), accumulator.f0);
            accumulator.f1.max(value.getTamperature(), accumulator.f1);
            return accumulator;
        }

        @Override
        public Tuple2<Double, Double> getResult(Tuple2<Double, Double> accumulator) {
            return accumulator;
        }

        @Override
        public Tuple2<Double, Double> merge(Tuple2<Double, Double> a, Tuple2<Double, Double> b) {
            return null;
        }
    }

    public static class Win extends ProcessWindowFunction<Tuple2<Double, Double>, MinMax, String, TimeWindow> {
        @Override
        public void process(String s, Context context, Iterable<Tuple2<Double, Double>> elements, Collector<MinMax> out) throws Exception {
            long start = context.window().getStart();
            long end = context.window().getEnd();
            Tuple2<Double, Double> element = null;
            for (Tuple2<Double, Double> e : elements) {
                if (element != null) return;
                element = e;
            }

            out.collect(new MinMax(s, element.f0, element.f1, "start:" + start + ",end:" + end));
        }
    }
}

File: flink-scala-core/src/main/scala/com/atguigu/flink/day3/MinMaxTempAggWindow.scala
package com.atguigu.flink.day3

import com.atguigu.flink.day2.{SensorReading, SensorSource}
import org.apache.flink.api.common.functions.AggregateFunction
import org.apache.flink.streaming.api.scala._
import org.apache.flink.streaming.api.scala.function.ProcessWindowFunction
import org.apache.flink.streaming.api.windowing.time.Time
import org.apache.flink.streaming.api.windowing.windows.TimeWindow
import org.apache.flink.util.Collector

object MinMaxTempAggWindow {

    def main(args: Array[String]): Unit = {
        val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
        env.setParallelism(1)

        env.addSource(new SensorSource)
                        .keyBy(_.id)
                        .timeWindow(Time.seconds(5))
                        .aggregate(new Agg,new Win)
                        .print()
        env.execute()
    }

    class Agg extends AggregateFunction[SensorReading,(Double,Double),(Double,Double)]{
        override def createAccumulator(): (Double, Double) = (0.0,0.0)

        override def add(value: SensorReading, accumulator: (Double, Double)): (Double, Double) = {
            (accumulator._1.min(value.temperature),accumulator._2.max(value.temperature))
        }

        override def getResult(accumulator: (Double, Double)): (Double, Double) = accumulator

        override def merge(a: (Double, Double), b: (Double, Double)): (Double, Double) = (a._1.min(b._1),a._2.max(b._2))
    }
    class Win extends ProcessWindowFunction[(Double,Double),MinMax,String,TimeWindow]{
        override def process(key: String, context: Context, elements: Iterable[(Double, Double)], out: Collector[MinMax]): Unit = {
            val start: Long = context.window.getStart
            val end: Long = context.window.getEnd
            out.collect(new MinMax(key,elements.head._1,elements.head._2,s"start:${start},end:${end}"))
        }
    }

    case class MinMax(key:String,min:Double,max:Double,out:String)
}
